File: app/agents/skilled.py
"""SkilledAgent: o agente flexível, dirigido por skills.

Substitui os antigos agentes rígidos por intenção (FAQ/Order/Support/Clarification).
Em vez de um classificador escolher UMA classe, este único agente recebe o
CONJUNTO de skills do agente (app.skills.enabled_skills_for) e deixa o LLM decidir
qual chamar, via function calling. É a generalização do loop que vivia no OrderAgent.

Fluxo de `execute`:

  1. Fast-path determinístico (0 token de LLM — preserva a economia):
     a. palavra-chave forte de escalonamento -> escalate_to_human direto;
     b. match RAG fortíssimo -> responde o chunk literal (atalho do antigo FAQ).
  2. Caso geral: loop de function-calling sobre as skills habilitadas. O modelo
     chama skills, recebe os resultados (JSON) e compõe a resposta em texto.
     - skill terminal (com direct_response, ex.: escalate) encerra o turno;
     - skill que pede handoff sem texto (ex.: reserve_stock) marca handoff e
       deixa o modelo redigir a confirmação.

O SDK do Groq é síncrono, então o loop inteiro roda em `asyncio.to_thread`.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import List, Optional

from app.agents.base import BaseAgent, build_messages, parse_handoff
from app.domain import AgentResult, ChatMessage
from app.llm import ToolUseFailedError, _tokens
from app.messages import DEGRADED_CATALOG, ORDER_CONFIRMED
from app.prompts import skilled_prompt
from app.skills import SkillContext, enabled_skills_for, invoke_skill, tool_schemas_for
from app.skills.knowledge import SHORTCUT_MAX_DISTANCE
from app.skills.support import ESCALATION_KEYWORDS
from app.textutil import word_set

log = logging.getLogger("blip-agent.skilled")

MAX_TOOL_ITERATIONS = 5  # teto de rodadas de tools por mensagem
TOOL_CALL_RETRIES = 2    # retries de uma chamada após tool call malformada


class SkilledAgent(BaseAgent):
    source = "llm"

    def __init__(self, agent_config, llm):
        super().__init__(agent_config, llm)
        self.skill_names = enabled_skills_for(agent_config)
        self.tools = tool_schemas_for(self.skill_names)
        self.ctx = SkillContext(agent=agent_config)

    def system_prompt(self, user_message: str) -> str:
        return skilled_prompt(self.agent, self.skill_names)

    async def execute(self, user_message: str, history: List[ChatMessage]) -> AgentResult:
        # Fast-path 1: escalonamento determinístico por palavra-chave (0 token).
        if "escalate_to_human" in self.skill_names and (word_set(user_message) & ESCALATION_KEYWORDS):
            log.info("Escalonamento determinístico (0 tokens).")
            res = invoke_skill("escalate_to_human",
                               {"reason": "Palavra-chave de escalonamento."}, self.ctx)
            return AgentResult(
                response=res.direct_response or "",
                should_handoff=True,
                handoff_reason=res.handoff_reason,
                source="support_escalation",
            )

        # Fast-path 2: atalho RAG fortíssimo (0 token de LLM; só um embedding).
        if "knowledge_search" in self.skill_names and self.agent.rag_enabled:
            shortcut = await asyncio.to_thread(self._rag_shortcut, user_message)
            if shortcut is not None:
                return shortcut

        # Caso geral: loop de function-calling sobre as skills habilitadas.
        return await asyncio.to_thread(self._run_loop, user_message, history)

    # --- fast-path RAG ----------------------------------------------------- #
    def _rag_shortcut(self, user_message: str) -> Optional[AgentResult]:
        """Reusa a skill knowledge_search; se o melhor match for fortíssimo,
        responde o chunk literal sem LLM (mesmo atalho do antigo FAQAgent)."""
        res = invoke_skill("knowledge_search", {"query": user_message}, self.ctx)
        results = (res.data or {}).get("results") or []
        if results and results[0]["score"] <= SHORTCUT_MAX_DISTANCE:
            top = results[0]
            log.info("RAG shortcut: source=%s dist=%.3f", top["source"], top["score"])
            return AgentResult(
                response=top["content"],
                source="faq_shortcut",
                rag_chunks_used=1,
                rag_sources=[top["source"]],
            )
        return None

    # --- loop de tools (bloqueante, roda em worker thread) ----------------- #
    def _run_loop(self, user_message: str, history: List[ChatMessage]) -> AgentResult:
        messages = build_messages(self.system_prompt(user_message), user_message, history)
        self._tokens_used = 0
        self._tools_called: list[str] = []
        self._sources: list[str] = []
        self._rag_chunks = 0
        self._pending_handoff = False
        self._handoff_reason: Optional[str] = None

        # Sem tools (caso raro): chat simples + parse de handoff pelo token.
        if not self.tools:
            text, tokens = self.llm.complete_sync(messages)
            self._tokens_used += tokens
            should, clean, reason = parse_handoff(text or "")
            return AgentResult(response=clean, should_handoff=should,
                               handoff_reason=reason, source="llm",
                               tokens_used=self._tokens_used)

        try:
            response = self._call(messages)
        except ToolUseFailedError:
            log.warning("Tool call malformada já na 1ª rodada; degradando.")
            return self._result(DEGRADED_CATALOG)

        iterations = 0
        while (response.choices[0].finish_reason == "tool_calls"
               and iterations < MAX_TOOL_ITERATIONS):
            iterations += 1
            messages.append(response.choices[0].message)  # pedido de tools do modelo

            terminal = self._handle_tool_calls(response.choices[0].message.tool_calls, messages)
            if terminal is not None:
                return terminal  # skill terminal (ex.: escalate) encerra o turno

            try:
                response = self._call(messages)
            except ToolUseFailedError:
                log.warning("Tool call malformada no meio do loop; finalizando em texto.")
                return self._result(self._finalize_in_text(messages))

        return self._result(response.choices[0].message.content or "")

    def _handle_tool_calls(self, tool_calls, messages: List[dict]) -> Optional[AgentResult]:
        """Executa cada skill pedida pelo modelo, devolve o resultado ao contexto
        e acumula metadados. Retorna um AgentResult se uma skill for TERMINAL
        (handoff com mensagem pronta), senão None (o loop continua)."""
        for tool_call in tool_calls:
            name = tool_call.function.name
            self._tools_called.append(name)
            try:
                args = json.loads(tool_call.function.arguments or "{}")
            except json.JSONDecodeError:
                args = {}

            res = invoke_skill(name, args, self.ctx)
            messages.append({
                "role": "tool",
                "tool_call_id": tool_call.id,
                "content": res.tool_payload(),
            })

            if name == "knowledge_search":
                self._rag_chunks += len((res.data or {}).get("results") or [])
            if res.sources:
                self._sources.extend(res.sources)

            if res.handoff:
                if res.direct_response:  # skill terminal -> encerra o turno já
                    return AgentResult(
                        response=res.direct_response,
                        should_handoff=True,
                        handoff_reason=res.handoff_reason,
                        source="support_escalation" if name == "escalate_to_human" else self.source,
                        tokens_used=self._tokens_used,
                        tools_called=self._tools_called,
                    )
                self._pending_handoff = True
                self._handoff_reason = res.handoff_reason
        return None

    def _call(self, messages: List[dict]):
        """complete_with_tools com retries para tool calls malformadas."""
        last_exc = None
        for attempt in range(TOOL_CALL_RETRIES + 1):
            try:
                response = self.llm.complete_with_tools(messages=messages, tools=self.tools)
                self._tokens_used += _tokens(response)
                return response
            except ToolUseFailedError as e:
                last_exc = e
                log.info("Retry %d/%d após tool_use_failed.", attempt + 1, TOOL_CALL_RETRIES)
        raise last_exc

    def _finalize_in_text(self, messages: List[ChatMessage]) -> str:
        """Pede uma resposta em texto puro com os dados de skills já coletados."""
        messages = messages + [{
            "role": "system",
            "content": ("Responda agora em texto natural, em português, usando os dados "
                        "das ferramentas já consultados. NÃO chame mais ferramentas."),
        }]
        try:
            text, tokens = self.llm.complete_sync(messages)
            self._tokens_used += tokens
            return text or DEGRADED_CATALOG
        except Exception as e:  # noqa: BLE001 — último recurso.
            log.warning("Finalização em texto falhou: %s", e)
            return DEGRADED_CATALOG

    def _result(self, text: str) -> AgentResult:
        text = (text or "").strip()
        should, clean, reason = parse_handoff(text)
        should = should or self._pending_handoff
        reason = reason or (self._handoff_reason if self._pending_handoff else None)
        if not clean:
            clean = ORDER_CONFIRMED if self._pending_handoff else DEGRADED_CATALOG
        # source informativo: RAG se a base foi consultada, senão llm.
        source = "llm_rag" if self._rag_chunks > 0 else "llm"
        log.info("skilled ok: tools=%s tokens=%d handoff=%s",
                 self._tools_called, self._tokens_used, should)
        return AgentResult(
            response=clean,
            should_handoff=should,
            handoff_reason=reason,
            source=source,
            tokens_used=self._tokens_used,
            tools_called=self._tools_called,
            rag_chunks_used=self._rag_chunks,
            rag_sources=sorted(set(self._sources)),
        )

File: src/server/agents/skilled.ts
/**
 * SkilledAgent: o agente flexível, dirigido por skills (porta `app/agents/skilled.py`).
 *
 * Recebe o CONJUNTO de skills do agente e deixa o LLM decidir qual chamar (function
 * calling). Um fast-path determinístico (0 token) preserva a economia:
 *   a. palavra-chave forte de escalonamento -> escalate_to_human direto.
 * Caso geral: loop de function-calling com teto e retries para tool_use_failed.
 */

import { agentResult, type AgentConfig, type AgentResult, type ChatMessage } from "../domain";
import { getLogger } from "../logging";
import { type ChatCompletion, type ChatMessageParam, type ChatTool, type LLMClient, type ToolCall, type ToolResult, ToolUseFailedError, tokensOf } from "../llm";
import { DEGRADED_CATALOG, ORDER_CONFIRMED } from "../messages";
import { skilledPrompt } from "../prompts";
import { sanitizeUntrusted } from "../security/sanitize";
import { makeSentinel, wrapToolData } from "../security/spotlight";
import { enabledSkillsFor, invokeSkill, type SkillContext, toolSchemasFor } from "../skills";
import { ESCALATION_KEYWORDS } from "../skills/support";
import { wordSet } from "../textutil";
import { BaseAgent, buildMessages, type ExecuteOptions, parseHandoff } from "./base";

const log = getLogger("blip-agent.skilled");

const MAX_TOOL_ITERATIONS = 5; // teto de rodadas de tools por mensagem
const TOOL_CALL_RETRIES = 2; // retries de uma chamada após tool call malformada

interface LoopState {
  tokensUsed: number;
  toolsCalled: string[];
  sources: string[];
  ragChunks: number;
  pendingHandoff: boolean;
  handoffReason: string | null;
}

interface KnowledgeData {
  results?: Array<{ source: string; content: string; score: number }>;
}

export class SkilledAgent extends BaseAgent {
  source = "llm";
  private readonly skillNames: string[];
  private readonly tools: ChatTool[];
  private readonly ctx: SkillContext;

  constructor(agentConfig: AgentConfig, llm: LLMClient) {
    super(agentConfig, llm);
    this.skillNames = enabledSkillsFor(agentConfig);
    this.tools = toolSchemasFor(this.skillNames);
    this.ctx = { agent: agentConfig };
  }

  systemPrompt(): string {
    // Contrato do BaseAgent. O fluxo real (runLoop) monta o prompt com o
    // sentinela do turno; este caminho não é usado em runtime pelo SkilledAgent.
    return skilledPrompt(this.agent, this.skillNames, makeSentinel());
  }

  async execute(userMessage: string, history: ChatMessage[], opts?: ExecuteOptions): Promise<AgentResult> {
    const hardened = opts?.hardened ?? false;

    // Fast-path 1: escalonamento determinístico por palavra-chave (0 token).
    if (this.skillNames.includes("escalate_to_human") && hasEscalationKeyword(userMessage)) {
      log.info("Escalonamento determinístico (0 tokens).");
      const res = await invokeSkill("escalate_to_human", { reason: "Palavra-chave de escalonamento." }, this.ctx);
      return agentResult({
        response: res.directResponse ?? "",
        shouldHandoff: true,
        handoffReason: res.handoffReason,
        source: "support_escalation",
      });
    }

    // Caso geral: loop de function-calling sobre as skills habilitadas.
    return this.runLoop(userMessage, history, hardened);
  }

  // --- loop de tools ------------------------------------------------------- //
  private async runLoop(userMessage: string, history: ChatMessage[], hardened: boolean): Promise<AgentResult> {
    const sentinel = makeSentinel();
    const systemPrompt = skilledPrompt(this.agent, this.skillNames, sentinel, hardened);
    const messages = buildMessages(systemPrompt, userMessage, history, sentinel);
    const state: LoopState = {
      tokensUsed: 0,
      toolsCalled: [],
      sources: [],
      ragChunks: 0,
      pendingHandoff: false,
      handoffReason: null,
    };

    // Sem tools (caso raro): chat simples + parse de handoff pelo token.
    if (this.tools.length === 0) {
      const [text, tokens] = await this.llm.complete(messages);
      state.tokensUsed += tokens;
      const { shouldHandoff, clean, reason } = parseHandoff(text || "");
      return agentResult({
        response: clean,
        shouldHandoff,
        handoffReason: reason,
        source: "llm",
        tokensUsed: state.tokensUsed,
      });
    }

    let response: ChatCompletion;
    try {
      response = await this.call(messages, state);
    } catch (e) {
      if (e instanceof ToolUseFailedError) {
        log.warn("Tool call malformada já na 1ª rodada; degradando.");
        return this.result(DEGRADED_CATALOG, state);
      }
      throw e;
    }

    let iterations = 0;
    while (response.stoppedForTools && iterations < MAX_TOOL_ITERATIONS) {
      iterations++;
      // Reecoa o turno do assistente que pediu as ferramentas (texto + tool calls).
      messages.push({ role: "assistant", content: response.text, toolCalls: response.toolCalls });

      const terminal = await this.handleToolCalls(response.toolCalls, messages, state, sentinel);
      if (terminal !== null) return terminal; // skill terminal (ex.: escalate)

      try {
        response = await this.call(messages, state);
      } catch (e) {
        if (e instanceof ToolUseFailedError) {
          log.warn("Tool call malformada no meio do loop; finalizando em texto.");
          return this.result(await this.finalizeInText(messages, state), state);
        }
        throw e;
      }
    }

    return this.result(response.text, state);
  }

  private async handleToolCalls(
    toolCalls: ToolCall[],
    messages: ChatMessageParam[],
    state: LoopState,
    sentinel: string,
  ): Promise<AgentResult | null> {
    const results: ToolResult[] = [];
    for (const tc of toolCalls) {
      const name = tc.name;
      state.toolsCalled.push(name);
      const args = tc.args ?? {}; // a Anthropic já entrega o input desserializado

      const res = await invokeSkill(name, args, this.ctx);
      // Saída de skill = DADO não confiável (chunk de RAG, catálogo externo):
      // sanitiza e envolve no bloco delimitado antes de devolver ao LLM.
      const payload = wrapToolData(sentinel, name, sanitizeUntrusted(res.toolPayload()));
      results.push({ toolCallId: tc.id, content: payload });

      if (name === "knowledge_search") {
        state.ragChunks += ((res.data as KnowledgeData | null)?.results ?? []).length;
      }
      if (res.sources.length > 0) state.sources.push(...res.sources);

      if (res.handoff) {
        if (res.directResponse) {
          // skill terminal -> encerra o turno já (sem continuar a conversa)
          return agentResult({
            response: res.directResponse,
            shouldHandoff: true,
            handoffReason: res.handoffReason,
            source: name === "escalate_to_human" ? "support_escalation" : this.source,
            tokensUsed: state.tokensUsed,
            toolsCalled: state.toolsCalled,
          });
        }
        state.pendingHandoff = true;
        state.handoffReason = res.handoffReason;
      }
    }
    // Todos os resultados desta rodada vão num único turno user (Messages API).
    messages.push({ role: "tool", results });
    return null;
  }

  /** completeWithTools com retries para tool calls malformadas. */
  private async call(messages: ChatMessageParam[], state: LoopState): Promise<ChatCompletion> {
    let lastExc: unknown = null;
    for (let attempt = 0; attempt <= TOOL_CALL_RETRIES; attempt++) {
      try {
        const response = await this.llm.completeWithTools(messages, this.tools);
        state.tokensUsed += tokensOf(response);
        return response;
      } catch (e) {
        if (e instanceof ToolUseFailedError) {
          lastExc = e;
          log.info(`Retry ${attempt + 1}/${TOOL_CALL_RETRIES} após tool_use_failed.`);
          continue;
        }
        throw e;
      }
    }
    throw lastExc;
  }

  private async finalizeInText(messages: ChatMessageParam[], state: LoopState): Promise<string> {
    const extended: ChatMessageParam[] = [
      ...messages,
      {
        role: "system",
        content:
          "Responda agora em texto natural, em português, usando os dados das " +
          "ferramentas já consultados. NÃO chame mais ferramentas.",
      },
    ];
    try {
      const [text, tokens] = await this.llm.complete(extended);
      state.tokensUsed += tokens;
      return text || DEGRADED_CATALOG;
    } catch (e) {
      log.warn("Finalização em texto falhou:", String(e));
      return DEGRADED_CATALOG;
    }
  }

  private result(text: string, state: LoopState): AgentResult {
    const { shouldHandoff, clean, reason } = parseHandoff((text || "").trim());
    const finalHandoff = shouldHandoff || state.pendingHandoff;
    const finalReason = reason ?? (state.pendingHandoff ? state.handoffReason : null);
    let finalClean = clean;
    if (!finalClean) finalClean = state.pendingHandoff ? ORDER_CONFIRMED : DEGRADED_CATALOG;
    const source = state.ragChunks > 0 ? "llm_rag" : "llm";
    log.info(`skilled ok: tools=${JSON.stringify(state.toolsCalled)} tokens=${state.tokensUsed} handoff=${finalHandoff}`);
    return agentResult({
      response: finalClean,
      shouldHandoff: finalHandoff,
      handoffReason: finalReason,
      source,
      tokensUsed: state.tokensUsed,
      toolsCalled: state.toolsCalled,
      ragChunksUsed: state.ragChunks,
      ragSources: [...new Set(state.sources)].sort(),
    });
  }
}

function hasEscalationKeyword(userMessage: string): boolean {
  for (const w of wordSet(userMessage)) {
    if (ESCALATION_KEYWORDS.has(w)) return true;
  }
  return false;
}
